Advanced tutorial - `Torchscript` image classifier (#395)
flojoy-ai/docs: docs/advanced-usage/pytorch-classifier.md

---
sidebar_position: 4
id: pytorch-classifier
title: Integrating Your Custom PyTorch Image Classifier
---

> In this guide, we will walk you through the process of leveraging the `TORCHSCRIPT_CLASSIFIER` node in Flojoy to integrate a custom image classification model. Specifically, we'll use a Colab notebook to train the `MobileNetV3` model on a cats and dogs dataset, export it as a `.torchscript` format, and then deploy it within the `TORCHSCRIPT_CLASSIFIER` node.

Reviewer: How about documenting TorchScript?
```suggestion
In this guide, we will walk you through the process of leveraging the `TORCHSCRIPT_CLASSIFIER` node in Flojoy to integrate a custom image classification model using [TorchScript](https://pytorch.org/docs/stable/jit.html). Specifically, we will use a Google Colab notebook to train the `MobileNetV3` model on a cats and dogs dataset, export it under the TorchScript format as `model.torchscript`, and then deploy it within the `TORCHSCRIPT_CLASSIFIER` node.
```

### 1. Train and Export Your Model

Start by accessing the provided Colab notebook which serves as an ideal foundation for training and exporting your model. This notebook is tailored for a cats vs. dogs dataset, but you can effortlessly modify it to suit different datasets with varying class numbers.

> - **Colab Notebook**: [Link](https://colab.research.google.com/drive/1HCj3i43tYt4CTLPHq4BqVC2HL0lAN2-X?usp=sharing)

Reviewer: Side-note: this is convenient, but this constraints Flojoy's documentation to a third-party service (Google Colab).

Author: Hmm great point, although Colab is a third-party service, I am confident the benefits outweigh the disadvantages, and here are a few reasons why

- Colab is here to stay in the ML ecosystem and remain freely accessible
- It provides a great reference implementation for people to compare with their own in-house implementation
- Allows anyone to quickly and instantly get started with a premade Cats and Dogs example
- Is more readable and interactive than a Python script
- Will feel very familiar to people who have used it in the past

Author: P.S: I added a disclaimer at the beginning of the tutorial to help drive the point home

Reviewer: Thanks!

I also think Colab is adapted and using it here is relevant for all the points you indicated. People might not have access to GPU and Colab is handy to access some.

I was unclear, what I meant was that we could indicate it up-front and your disclaimer does it.

To use the notebook:

1. Open the notebook link.
2. Run all cells sequentially.
3. Upon completion, you'll find the `model.torchscript` file ready for download in the left sidebar, as illustrated here:

![image](../../static/img/advanced_tutorials/TORCHSCRIPT_TUTORIAL_SIDEBAR.png)

### 2. Deploy the Model in `TORCHSCRIPT_CLASSIFIER`

To utilize a trained model within the `TORCHSCRIPT_CLASSIFIER` node, ensure you have:

- The exported `.torchscript` model file.
- A `.csv` file mapping class names to class IDs.

For instance, your `csv` file for a cats vs. dogs classifier should look like this:

```
class_name
"cat"
"dog"
```

> Now, with these files at hand, you're all set to integrate your custom PyTorch Image Classifier into Flojoy using the `TORCHSCRIPT_CLASSIFIER` node.

Reviewer: Nitpick: how formal do we want the documentation to be?
```suggestion
Now, with these files at hand, you are all set to integrate your custom PyTorch Image Classifier into Flojoy using the `TORCHSCRIPT_CLASSIFIER` node.
```

> ### 3. [Optional] Train the model on your own data

Reviewer: Could you explain how to specify the ground truth for training model?

If you'd like to train the model on your own data, your dataset needs to be compatible with the PyTorch `ImageFolder` dataset format. This essentially means that your images need to be structured in a directory hierarchy such as:

```
dataset/
├── class_1
│   ├── img1.jpg
│   ├── img2.jpg
│   ├── ...
├── class_2
│   ├── img1.jpg
│   ├── img2.jpg
│   ├── ...
├── ...
```

For more information on the `ImageFolder` dataset format, refer to the PyTorch documentation [here](https://pytorch.org/vision/stable/generated/torchvision.datasets.ImageFolder.html#torchvision.datasets.ImageFolder).

> You will need to curate both (a) training and (b) test sets with the format described above, and compress them together in a zip file containing a `training_set` folder and a `test_set` folder.

Reviewer: To me, we must explain how we evaluation of models is performed. For now, there is an ambiguity in the ways models are trained and evaluated. In the notebook, `val_loader`, the Data Loader for validation, use the `test_set` ImageFolder.

How about renaming `test_set` as `validation_set`?

```suggestion
You will need to create an zip archive consisting of two folders `training_set` and `validation_set`.

Both `training_set` and `validation_set` have to comply with the [`torchvision.datasets.ImageFolder`](https://pytorch.org/vision/stable/generated/torchvision.datasets.ImageFolder.html#torchvision.datasets.ImageFolder) format.
```

Author: Valid point, thanks for nuancing this

```
dataset.zip/
├── training_set/
│   ├── class_1
│   │   ├── img1.jpg
│   │   ├── img2.jpg
│   │   ├── ...
│   ├── class_2
│   │   ├── img1.jpg
│   │   ├── img2.jpg
│   │   ├── ...
│   ├── ...

> ├── test_set/

Reviewer: Following the suggestion made above:
```suggestion
├── validation_set/
```

---

sidebar_position: 4
id: pytorch-classifier
title: Integrating Your Custom PyTorch Image Classifier

---

In this guide, we will walk you through the process of leveraging the [`TORCHSCRIPT_CLASSIFIER`](../nodes/AI_ML/CLASSIFICATION/TORCHSCRIPT_CLASSIFIER/TORCHSCRIPT_CLASSIFIER.md) node in Flojoy to integrate a custom image classification model. 

> Solely for illustration purposes, we will use a companion [Colab](https://research.google.com/colaboratory/) notebook to train the `MobileNetV3` model on a cats and dogs dataset, export it as a `.torchscript` format, and then deploy it within the `TORCHSCRIPT_CLASSIFIER` node. Flojoy users do not have to use the same notebook, nor even use Colab at all, though it is a great place to help users get started with a reference implementation example.

Reviewer: Last nitpick from me!
```suggestion
Solely for illustration purposes, we will use a companion [Colab](https://research.google.com/colaboratory/) notebook to train a [`MobileNetV3`](https://arxiv.org/abs/1905.02244) model on a cats and dogs dataset, export it as a `.torchscript` format, and then deploy it within the `TORCHSCRIPT_CLASSIFIER` node. Flojoy users do not have to use the same notebook, nor even use Colab at all, though it is a great place to help users get started with a reference implementation example.
```

```
dataset/
├── class_1
│   ├── img1.jpg
│   ├── img2.jpg
│   ├── ...
├── class_2
│   ├── img1.jpg
│   ├── img2.jpg
│   ├── ...
├── ...
```

#### Instructions:

1. Curate your `training_set` and `validation_set` folders separately, and ensure that both folders respect the `ImageFolder` format (described above).

2. Compress both `training_set` and `validation_set` into a `.zip` folder that you may call `datasets.zip` for example. 
    
    _Example_

    ```
    datasets.zip/
    ├── training_set/
    │   ├── ... 
    ├── validation_set/
    │   ├── ...
    ```

3. Drag and drop `datasets.zip` into Google Colab's left file sidebar, and wait until the upload completes.

4. If `training_set` and `validation_set` folders already exist in your Colab file tree, make sure to delete them before you proceed.

5. In the "Download dataset" section, comment (add `#` at the beginning of each line) the two original lines that download and unzip `cats_and_dogs.zip` and uncomment the `!unzip datasets.zip` command
```
# The two lines below download and unzip 'cats_and_dogs.zip'
#!gdown  https://drive.google.com/uc?id=1TIme-TSfRFwVvBVaGymV4-9CX4voThIg
#!unzip cats_and_dogs.zip

> # [Advanced tutorial - Optional section 3] Comment the 2 lines above and uncomment the line below to use your own dataset 

Reviewer: To avoid this line from being wrapped around.
```suggestion
# [Advanced tutorial - Optional section 3]
# Comment the 2 lines above and uncomment the line below to use your own dataset 
```
